aiven/client/cli.py

```diff
@@ -101,6 +103,82 @@ def add_args(self, parser):
+        for prop, spec in sorted(obj_def.get("properties", {}).items()):
+            full_name = prop if not prefix else (prefix + "." + prop)
+            types = spec["type"]
+            if not isinstance(types, list):
+                types = [types]
+            # "object" or ["object", "null"]
+            if "object" in types:
+                opts.update(self.collect_user_config_options(spec, prefix=full_name))
+                if "null" in types:
+                    # allow removing user config option
+                    opts[full_name] = {
+                        "title": "Remove {}".format(prop),
+                        "type": "null",
+                    }
+            else:
+                opts[full_name] = spec
+        for spec in sorted(obj_def.get("patternProperties", {}).values()):
+            full_name = "KEY" if not prefix else (prefix + ".KEY")
+            if spec["type"] == "object":
+                opts.update(self.collect_user_config_options(spec, prefix=full_name))
+            else:
+                opts[full_name] = spec
+        return opts
+
+    def create_user_config(self, user_config_schema):
+        """Convert a list of ["foo.bar='baz'"] to {"foo": {"bar": "baz"}}"""
+        if not self.args.user_config and not self.args.user_option_remove:
+            return {}
+
+        options = self.collect_user_config_options(user_config_schema)
+        user_config = {}
+        for key_value in self.args.user_config:
+            try:
+                key, value = key_value.split("=", 1)
+            except ValueError:
+                raise argx.UserError("Invalid config value: {!r}, expected '<KEY>[.<SUBKEY>]=<JSON_VALUE>'"
+                                     .format(key_value))
+
+            opt_schema = options.get(key)
+            if not opt_schema:
+                # Exact key not found, try generic one
+                generic_key = ".".join(key.split(".")[:-1] + ["KEY"])
+                opt_schema = options.get(generic_key)
+
+            if not opt_schema:
+                raise argx.UserError("Unsupported option {!r}, available options: {}"
+                                     .format(key, ", ".join(options) or "none"))
+
+            try:
+                value = convert_str_to_value(opt_schema, value)
+            except ValueError as ex:
+                raise argx.UserError("Invalid value {!r}: {}".format(key_value, ex))
+
+            conf = user_config
+            parts = key.split(".", 1)
+            for part in parts[:-1]:
+                conf.setdefault(part, {})
+                conf = conf[part]
+
+            conf[parts[-1]] = value
+
+        for opt in self.args.user_option_remove:
+            opt_schema = options.get(opt)
+            if not opt_schema:
+                raise argx.UserError("Unsupported option {!r}, available options: {}"
+                                     .format(opt, ", ".join(options) or "none"))
+
+            if opt_schema["type"] != "null":
```

Comment:
Shouldn't this rather do `if "null" not in opt_schema["type"]:` ? Current implementation only works for type object, not for e.g. string that allows both null and the string.